site/_posts/2023-07-11-get-more-value-out-of-fusionauth.md

```diff
@@ -0,0 +1,321 @@
+
+```cs
+//For Group object
+public class Group
+{
+    public string ID;
+    public List<GroupUser> Users;
+}
+
+public class GroupUser
+{
+    [JsonProperty(PropertyName = "userId")]
+    public string UserID;
+}
+```
+<br>
+
+### **Input**
+As with creating the user, we need to take the input from the application and assign the values to our Group objects. We can also validate any data we want before actually adding the user to the group. After we are satisfied the user should be added, we can submit the information to the method that will submit our request to the FusionAuth server.
+
+```cs
+private async void btnAddUserToGroup_Click(object sender, EventArgs e)
+{
+    //FusionAuthClient is a class that contains the logic to call  the FusionAuth APIs
+    var faClient = new FusionAuthClient();
+
+    List<GroupUser> users = new List<GroupUser>()
+    {
+        new GroupUser() 
+        {
+            UserID = txtUserIDToAddToGroup.Text
+        }
+    };
+
+    Group group = new Group();
+
+    group.ID = txtGroupIDToAddTo.Text;
+    group.Users = users;
+
+    List<Group> groups = new List<Group>();
+    groups.Add(group);
+
+    //perform custom validation logic here
+    if (IsGroupValid(group))
+    {
+        ReturnValue addUserToGroupReturnValue = await faClient.AddUserToGroup(groups);
+        if (addUserToGroupReturnValue.success == true)
+        {
+            LogResults($"User has been added to the group: {addUserToGroupReturnValue.result}");
+        }
+        else
+        {
+            LogResults($"Adding the user to the group has failed. {addUserToGroupReturnValue.result}");
+        }
+    }
+    else
+    {
+        LogResults($"User and/or group data provided is invalid");
+    }
+}
+```
+
+<br>
+
+### **Add the User to the Group**
+Because the JSON format expected in the request body is not readily serializable given the Group objects we have created, we will have to perform a little wizardry to manipulate the Group objects. While this does not have to do with FusionAuth specifically, I have abstracted the code for that in the FormatGroupUserJSON function. The code for that will be in the project. Once the API call has returned, you can retrieve whichever values you see as relevant and pass it back in the result object.
+
+```cs
+public async Task<ReturnValue> AddUserToGroup(List<Group> groups)
+{
+    //returns HttpClient with correct setting for instance
+    HttpClient httpClient = GetClient();
+
+    //initialize the return value 
+    ReturnValue returnVal = new ReturnValue()
+    {
+        success = false
+    };
+
+    try
+    {
+        //set the endpoint for creating a user
+        string apiURL = $"/api/group/member";
+
+        var payload = FormatGroupUserJSON(groups);
+        var content = new StringContent(payload, Encoding.UTF8, "application/json");
+
+        HttpResponseMessage response = await httpClient.PostAsync($"{Constants.BASE_URL}{apiURL}", content);
```

Comment:
Again, have qs about this actually working without the API key. What am I missing?

Reply:
the httpClient is set with the call to the FusionAuthClient.GetClient().  That contains the droid you are looking for:
```
private HttpClient GetClient()
        {
            HttpClient httpClient = new HttpClient();
            //The Constants class contains the infromation needed for the URL and API KEY.
            //i.e BASE_URL = "http://localhost:9011";
            //i.e. API_KEY = "XLE1_6EawwxDsnM9x81uOqbxaOyO759aMXM25_edSThyPCHPsBT_8M9K";
            httpClient.BaseAddress = new Uri(Constants.BASE_URL);
            httpClient.DefaultRequestHeaders.Add("Authorization", Constants.API_KEY);
            return httpClient;
        }
```

I will look at ways to make this more clear.

Reply:
Great. Also, please have the API key pulled from a settings file or environment variable.

Reply:
updated app and done.